content/blog/2022-05-03-e2e-cv-pt1.md

```diff
@@ -0,0 +1,316 @@
+
+For security reasons, you'll need to set up all keys and tokens through
+[GitHub secrets](https://docs.github.com/en/actions/security-guides/encrypted-secrets).
+You'll also need to change the remote location (and its name) in the
+[DVC config](https://github.com/iterative/magnetic-tiles-defect/blob/main/.dvc/config)
+file for versioning data and other artifacts.
+
+### Proof-of-Concept in Jupyter Notebooks
+
+A typical ML project would start with data collection and/or labeling, but we
+are skipping all this hard work because it was done for us by the researchers
+who published the dataset.
+
+We'll get right to the exciting part of training CV models in Jupyter notebooks
+which you can find
+[here](https://github.com/iterative/magnetic-tiles-defect/tree/main/notebooks).
+In short, there we have three notebooks:
+
+1. [`1_ProcessData.ipynb`](https://github.com/iterative/magnetic-tiles-defect/blob/main/notebooks/1_ProcessData.ipynb) -
+   downloads, processes, and organizes the data for easy loading into the
+   training process later
+2. [`2_TrainSegmentationModel.ipynb`](https://github.com/iterative/magnetic-tiles-defect/blob/main/notebooks/2_TrainSegmentationModel.ipynb) -
+   uses [`fastai`](https://github.com/fastai/fastai) Deep Learning framework to
+   train an image segmentation model
+3. [`3_Evaluate.ipynb`](https://github.com/iterative/magnetic-tiles-defect/blob/main/notebooks/3_Evaluate.ipynb) -
+   computes model performance on the test dataset
+
+Jupyter Notebook is by far the most popular tool for quick exploratory work when
+it comes to data analysis and modeling. However, it's not without
+[its own limitations](https://www.youtube.com/watch?v=7jiPeIFXb6U). One of the
+biggest issues of Jupyter is that it has no guardrails to ensure
+reproducibility, e.g. hidden states of variables and objects as well as the
+possibility to run cells out of order. While there are several projects that
+attempt to alleviate some of these issues (notably,
+[`nodebook`](https://github.com/stitchfix/nodebook),
+[`papermill`](https://github.com/nteract/papermill),
+[`nbdime`](https://github.com/jupyter/nbdime),
+[`nbval`](https://github.com/computationalmodelling/nbval),
+[`nbstripout`](https://github.com/kynan/nbstripout), and
+[`nbQA`](https://github.com/nbQA-dev/nbQA)), they don’t solve them completely.
+
+That's where the concepts of data versioning and ML pipelines come in.
+
+### Data Versioning
+
+Here, we'll briefly touch on DVC's functionality to version data and other
+artifacts (e.g. model weights).
+
+In most ML projects, training data changes gradually over time as new training
+instances (images in our case) get added while older ones might be removed.
+Simply creating snapshots of our training data at the time of training (e.g.
+labeling data directories with dates) quickly becomes unsustainable since these
+snapshots will contain many duplicates.
```

Comment:
Might also want to mention organization as another benefit. Tracking which date/dir was used for each model can become a mess.

Reply:
added a clarifying sentence